app/workers/delete_object_hierarchy_worker.rb

```diff
@@ -2,148 +2,177 @@
+    else
+      raise ArgumentError, "Invalid entry specification: #{entry}"
+    end
+  rescue ActiveRecord::RecordNotFound => exception
+    Rails.logger.warn "#{self.class} skipping object, maybe something else already deleted it: #{exception.message}"
+    []
+  rescue NameError
+    raise DoNotRetryError, "seems like unexpectedly broken delete hierarchy entry: #{entry}"
   end
 
-  class ReflectionDestroyer
-
-    def initialize(main_object, reflection, caller_worker_hierarchy)
-      @main_object = main_object
-      @reflection = reflection
-      @caller_worker_hierarchy = caller_worker_hierarchy
+  # @return a single associated object for deletion or nil if non in the association
+  def handle_association(ar_object, association, hierarchy_association_string)
+    reflection = ar_object.class.reflect_on_association(association)
+    case reflection.macro
+    when :has_many
+      # here we keep original hierarchy entry if we still find an associated object
+      dependent = ar_object.public_send(association).take
+      dependent ? [hierarchy_association_string, *hierarchy_entries_for(dependent)] : []
+    when :has_one
+      # maximum of one associated so we never keep the original hierarchy entry
+      hierarchy_entries_for ar_object.public_send(association)
+    else
+      raise ArgumentError, "Cannot handle association #{ar_object}:#{association} type #{reflection.macro}"
     end
+  end
 
-    def destroy_later
-      reflection.many? ? destroy_has_many_association : destroy_has_one_association
-    end
+  def compatibility(_object, caller_worker_hierarchy = [], _background_destroy_method = 'destroy')
+    # maybe requeue first object from hierarchy would be adequate and uniqueness should deduplicate jobs
+    hierarchy_root = Array(caller_worker_hierarchy).first
+    return _object && self.class.delete_later(_object) unless hierarchy_root
 
-    attr_reader :main_object, :reflection, :caller_worker_hierarchy
+    object_class, id = hierarchy_root.match(/Hierarchy-([a-zA-Z0-9_]+)-([\d*]+)/).captures
 
-    private
+    raise DoNotRetryError, "background deletion cannot handle #{hierarchy_root}" unless object_class && id
 
-    def destroy_has_many_association
-      main_object.public_send("#{reflection.name.to_s.singularize}_ids").each do |associated_object_id|
-        associated_object = reflection.class_name.constantize.new
-        associated_object.id = associated_object_id
-        delete_associated_object_later(associated_object)
-      end
-    rescue ActiveRecord::UnknownPrimaryKey => exception
-      Rails.logger.info "DeleteObjectHierarchyWorker#perform raised #{exception.class} with message #{exception.message}"
-    end
+    self.class.delete_later object_class.constantize.find(id.to_i)
+  rescue ActiveRecord::RecordNotFound => exception
+    Rails.logger.warn "#{self.class} skipping object, maybe something else already deleted it: #{exception.message}"
```

Comment:
Is it OK to not return anything in this case?

Reply:
We return `[]` which means just drop the entry we couldn't find an object for. The main loop will continue handling the next entry. `#handle_hierarchy_entry` is meant to return an array.

Reply:
But the rescue block doesn't return `[]`, it just logs a warning, it returns `nil` then.

Reply:
ah, this is the compatibility mode. Yes, it returns whatever and there is nothing queued further.